Adding version-conditional context string support

baentsch opened the pull request:
This also acts as a test for the new `liboqs` version define feature by conditionally enabling context string support during signature operations for plain PQ algorithms supporting that feature and in openssl versions also supporting that feature.

SWilson4 commented on oqsprov/oqs_sig.c:
```diff
@@ -1176,6 +1196,7 @@ static void oqs_sig_freectx(void *vpoqs_sigctx) {
     OPENSSL_free(ctx->aid);
     ctx->aid = NULL;
     ctx->aid_len = 0;
+    OPENSSL_free(ctx->context_string);
```
```suggestion
    OPENSSL_free(ctx->context_string);
    ctx->context_string = NULL;
    ctx->context_string_length = 0;
```
Kind of a trivial point, but it's done for the other struct members, so may as well be consistent.

baentsch replied:
I was briefly tempted to remove the other pretty pointless lines, but so what... See latest commit.

SWilson4 commented on oqsprov/oqs_sig.c:
```diff
@@ -1299,6 +1320,17 @@ static int oqs_sig_set_ctx_params(void *vpoqs_sigctx,
+    if (p != NULL) {
+        if (!OSSL_PARAM_get_octet_string(
+                p, &poqs_sigctx->context_string, 0,
+                &(poqs_sigctx->context_string_length))) {
```
```suggestion
                &poqs_sigctx->context_string_length)) {
```
The parens here aren't necessary, which isn't a problem _per se_, but it's a little confusing when they're included here but not on the line above.

baentsch replied:
Removed as part of the commit adding verify_with_ctx so this is then functionally complete and no-one can complain about the in-symmetry.

SWilson4 reviewed (approved):
LGTM! Thanks, Michael. I left a couple of picky style-related comments; feel free to ignore them if you wish.

SWilson4 commented:
> @SWilson4 Did you provide the tool "format_code.sh" running the docker image with the same `clang-format` version that CI uses? It now seems to fail: I just re-ran it and it confirms "no change" while CI does detect a change. Ideas what's wrong here?

It looks to me like the coding style tests are passing in the Ubuntu 24 image (which is what that script uses). However, the failing tests also seem to be running clang-format in other images (Ubuntu 22 / Alpine) :(

baentsch commented:
>  However, the failing tests also seem to be running clang-format in other images (Ubuntu 22 / Alpine) :(

Hmpf; OK, just saw https://github.com/open-quantum-safe/oqs-provider/blob/98ec7fc3e1f194b9df4760eccc1ca2cf2c0dc16b/.github/workflows/linux.yml#L46-L52

That's of course an invitation to fail -- interesting that it worked until today :-/. Objections to removing these style checks from these platforms?

SWilson4 commented:
> That's of course an invitation to fail -- interesting that it worked until today :-/. Objections to removing these style checks from these platforms?

None from me—especially since those tests [only run if the style checks pass on Ubuntu 24](https://github.com/open-quantum-safe/oqs-provider/blob/98ec7fc3e1f194b9df4760eccc1ca2cf2c0dc16b/.github/workflows/linux.yml#L37).

baentsch commented:
OK -- then what about putting these "nothing (in terms of style) changes on regenerate" tests also into the coding style check script? The re-generate and run tests on different platforms do make sense to retain.

SWilson4 commented:
> OK -- then what about putting these "nothing (in terms of style) changes on regenerate" tests also into the coding style check script? The re-generate and run tests on different platforms do make sense to retain (just no coding style check then there).

The purpose is to check idempotence under the operation of "regenerate and reformat", correct? (Similar to the `copy_from_upstream` check in liboqs.) If so, makes sense to me, and I agree that it makes sense to retain the regenerate -> run tests flow on different platforms.

I do wonder if there's value in also running the tests _without_ regenerating on different platforms. Previously we had assurance that no code differences arose as part of the regeneration process, so it didn't matter. I'm thinking of the following scenario:
1. There's a bug specific to Alpine (say) present in the code as available in GitHub.
2. The bug is silently fixed by regenerating the code on Alpine. CI doesn't catch the change since it now ignores code changes after regeneration.
3. Since we only run tests after regeneration, the bug is never detected.

Does this seem far-fetched to you, or is it something we need to worry about?

baentsch commented:
> Does this seem far-fetched to you, or is it something we need to worry about?

No, it doesn't. My approach would have been to just remove the `git diff` test from where it is right now and add a "regenerate-and-test-diff" test to the style check job using the ubuntu-latest image.

baentsch commented:
> No, it doesn't. My approach would have been to just remove the `git diff` test from where it is right now and add a "regenerate-and-test-diff" test to the style check job using the ubuntu-latest image.

So now implemented -- addressing all your concerns @SWilson4 ?

SWilson4 commented:
> So now implemented -- addressing all your concerns @SWilson4 ?

Yes, I'm happy. Thanks @baentsch!
